File: java/0094-binary-tree-inorder-traversal.java
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        //iterative
        List<Integer> res = new ArrayList<>();
        if(root == null) return res;

        Stack<TreeNode> stack = new Stack<>();
        TreeNode node = root;
        while(!stack.isEmpty() || node != null) {
            if(node != null) {
                stack.push(node);
                node = node.left;
            } else {
                node = stack.pop();
                res.add(node.val);
                node = node.right;
            }
        }

        return res;

File: kotlin/0094-binary-tree-inorder-traversal.kt
class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val res = ArrayList<Int>()
        
        fun inOrder(node: TreeNode?) {
            node?: return
            inOrder(node.left)
            res.add(node.`val`)
            inOrder(node.right)
        }
        
        inOrder(root)
        return res
    }
}
